server/cp/ultrad.py

```diff
@@ -0,0 +1,135 @@
+
+
+def upload_document(item):
+    payload = {
+        'lang': {
+            'fromLang': 'en',
+            'toLang': 'fr',
+        },
+        'name': item['headline'],
+        'state': 'new',
+        'text': {
+            'original': get_text(item['body_html']),
+        },
+    }
+
+    try:
+        resp = requests.post(ULTRAD_URL, json=payload, headers=get_headers(), timeout=ULTRAD_TIMEOUT)
+        resp.raise_for_status()
+    except requests.HTTPError:
+        logger.exception('got error on post %s', item['guid'])
+        raise UltradException()
+
+    data = get_json(resp)
+    return data['_id']
+
+
+def get_document(ultrad_id):
+    url = urljoin(ULTRAD_URL, ultrad_id)
+    try:
+        resp = requests.get(url, headers=get_headers(), timeout=ULTRAD_TIMEOUT)
+        resp.raise_for_status()
+    except requests.HTTPError:
+        logger.exception('got error on get %s', ultrad_id)
+        raise UltradException()
+    return get_json(resp)
+
+
+def get_json(resp):
+    try:
+        return resp.json()
+    except ValueError:
+        logger.error('error when parsing ultrad response "%s"', resp.text)
+        raise UltradException()
+
+
+@celery.task(soft_time_limit=300)
+def sync():
+    lock_name = 'ultrad'
+    if not lock(lock_name):
+        logger.info('lock taken %s', lock_name)
+        return
+    try:
+        desk = get_resource_service('desks').find_one(req=None, name=app.config['ULTRAD_DESK'])
+        if not desk:
+            logger.warn('ultrad desk not found name=%s', app.config['ULTRAD_DESK'])
+            return
+        todo_stage = get_resource_service('stages').find_one(req=None, desk=desk['_id'],
+                                                             name=app.config['ULTRAD_TODO_STAGE'])
+        done_stage = get_resource_service('stages').find_one(req=None, desk=desk['_id'],
+                                                             name=app.config['ULTRAD_DONE_STAGE'])
+        if not todo_stage:
+            logger.warn('ultrad todo stage is missing name=%s', app.config['ULTRAD_TODO_STAGE'])
+            return
+        if not done_stage:
+            logger.warn('ultrad done stage is missing name=%s', app.config['ULTRAD_DONE_STAGE'])
+            return
+        lookup = {'task.stage': todo_stage['_id']}
+        items = list(get_resource_service('archive').get(req=None, lookup=lookup))
+        logger.info('checking %d items on ultrad', len(items))
+        for item in items:
+            if not touch(lock_name, expire=300):
+                logger.warn('lost lock %s', lock_name)
+                break
+            if item.get('lock_user') and item.get('lock_session'):
+                logger.info('skipping locked item guid=%s', item['guid'])
+                continue
+            try:
+                ultrad_id = item['extra'][ULTRAD_ID]
+                ultrad_doc = get_document(ultrad_id)
+            except (KeyError, UltradException):
+                logger.exception('wtf')
```

Comment:
please log the whole exception (`wtf` is not really helping, even if funny :laughing: )